### PR TITLE
Add Figure.hlines for plotting horizontal lines

### PR DESCRIPTION
**Pull request by michaelgrund**
**Description of proposed changes**

This PR implements the `Figure.hlines` method to plot horizontal line(s). This is a high-level wrapper of `Figure.plot`, and is inspired by [`matplotlib.pyplot.hlines`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.hlines.html).

Address #670.

**Features**

- Plot one or multiple lines (specified by `y`)
- The X-limits of lines can be set by `xmin`/`xmax`, and then default to the x-limits of the current plot (i.e., `fig.region[0:2]`)
- Support Cartesian coordinates (as straight lines), geographic projection (as lines along constant latitude), and polar projection (as lines along constant radius).
- The line's pen attribute can be set via `pen`
- A legend entry can be added via `label`
- Parameters `no_clip`/~~`transparency`~~/`perspective` are also supported

**Limitations**

- When plotting multiple lines, only one single `pen`/`label` is allowed. It's possible to allow different *pen* or *label* (by passing a sequence of *pen* or *label*), but I feel it makes the codes too complicated. Users who need different pens/labels can call `Figure.hlines` multiple times.
- When plottting multiple lines, we call `Figure.plot` multiple times in a loop. As mentioned in https://github.com/GenericMappingTools/pygmt/pull/923#issuecomment-780972871, it's possible to use `np.nan` as separators for each line, then we can call `Figure.plot` once. However, GMT will complain about `NaN` records when plotting lines in geographic projections because GMT needs to resample the lines before plotting.
- When plottling lines in geographic or polar projections, just passing two points is not enough, since GMT defaults to plot the minor great-circle arcs. As shown below, we need to at least 4 points for geographic projections.
```
import pygmt
import numpy as np

fig = pygmt.Figure()
fig.basemap(region="d", projection="H15c", frame=True)

# Passing 2, 3 or 4 data points to plot horizontal lines.
for npoints in [2, 3, 4]:
    fig.plot(x=np.linspace(0, 360, npoints), y=[npoints * 10] * npoints, pen="1p,red", straight_line="m")
fig.show()
```
- `npoints=2`: Line is not plotted, because in geographic projections, longitude is periodic, so longitude 0=360.
- `npoints=3`: i.e., passing `x=[0, 180, 360]`, only one _minor_ arc is drawn
- `npoints=4`: i.e., passing `x=[0, 90, 180, 360]`, works
hlines-resampling

**Example**

Here is an example showing how it works in Cartesian/polar/geographic projections. See the tests for more examples.

```python
import pygmt

fig = pygmt.Figure()

# Cartesian projection
fig.basemap(region=[0, 10, 0, 10], projection="X10c/10c", frame=True)
fig.hlines(y=1, pen="1p,black", label="Line 1")
fig.hlines(y=2, xmin=2, xmax=8, pen="1p,red,-", label="Line 2")
fig.hlines(y=[3, 4], xmin=3, xmax=7, pen="1p,black,.", label="Line 3")
fig.hlines(y=[5, 6], xmin=4, xmax=9, pen="1p,red", label="Line 4")
fig.hlines(y=[7, 8], xmin=[0, 1], xmax=[7, 8], pen="1p,blue", label="Line 5")
fig.legend()

fig.shift_origin(xshift="w+2c")
# Geographic 
fig.basemap(region="g", projection="R15c", frame=True)
fig.hlines(50, xmin=0, xmax=360, pen="1p")

fig.shift_origin(xshift="w+2c")
# Polar
fig.basemap(region=[0, 360, 0, 1], projection="P10c", frame=True)
fig.hlines(0.5, xmin=0, xmax=80, pen="1p")
fig.show()
```

hlines

**Preview**: https://pygmt-dev--923.org.readthedocs.build/en/923/api/generated/pygmt.Figure.hlines.html

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [x] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash command is:

- `/format`: automatically format and lint the code

### CODE REVIEW

**Comment by seisman** [comment]
@michaelgrund Nice work! I just changed the PR title so that we can focus on one feature `Figure.hlines` in this PR.

**Comment by seisman** [comment]
> Furthermore, currently the whole procedure is performed in a loop since it's not possible to pass multi-segment lines to the plot function so far (except using an input file).

It's possible to use `np.NaN` record to separate segments:
```
import numpy as np
import pygmt

x = np.array([2, 4, 4, np.NaN, 5, 6])
y = np.array([3, 5, 7, np.NaN, 4, 2])

fig = pygmt.Figure()
fig.basemap(region=[0, 10, 0, 10], frame=True)
fig.plot(x, y, pen="2p")
fig.show()
```
but I'm not sure if other features (lines with different colors) are supported.

**Comment by michaelgrund** [comment]
> 
> 
> > Furthermore, currently the whole procedure is performed in a loop since it's not possible to pass multi-segment lines to the plot function so far (except using an input file).
> 
> It's possible to use `np.NaN` record to separate segments:
> 
> ```
> import numpy as np
> import pygmt
> 
> x = np.array([2, 4, 4, np.NaN, 5, 6])
> y = np.array([3, 5, 7, np.NaN, 4, 2])
> 
> fig = pygmt.Figure()
> fig.basemap(region=[0, 10, 0, 10], frame=True)
> fig.plot(x, y, pen="2p")
> fig.show()
> ```
> 
> but I'm not sure if other features (lines with different colors) are supported.

Ok will check that and, if possible, adjust the code.

File: pygmt/src/hlines.py
Comment (seisman) [inline pygmt/src/hlines.py:32]: IMHO, we should keep the function simple, and do not add aliases like `frame`, `projection`, `region`, `timestamp`, `xshift`, `yshift`, `zvalue`, `cmap`.

For people who want coloring lines based on z values (`zvalue` and `cmap`), they should use `Figure.plot()` for more features.

Comment (michaelgrund) [reply pygmt/src/hlines.py:32]: Ok, would be fine for me @seisman ! Means it should be only possible to plot lines with colors defined in `pen` and all pepole who want to plot advanced and fancier stuff should take a look at `Figure.plot()`?

Comment (seisman) [reply pygmt/src/hlines.py:32]: Yes.

**Comment by michaelgrund** [comment]
> 
> 
> > Furthermore, currently the whole procedure is performed in a loop since it's not possible to pass multi-segment lines to the plot function so far (except using an input file).
> 
> It's possible to use `np.NaN` record to separate segments:
> 
> ```
> import numpy as np
> import pygmt
> 
> x = np.array([2, 4, 4, np.NaN, 5, 6])
> y = np.array([3, 5, 7, np.NaN, 4, 2])
> 
> fig = pygmt.Figure()
> fig.basemap(region=[0, 10, 0, 10], frame=True)
> fig.plot(x, y, pen="2p")
> fig.show()
> ```
> 
> but I'm not sure if other features (lines with different colors) are supported.

I made some adjustments, however, as you mentioned @seisman it's not possible to put different linestyles or colors into ``plot`` separated by ``np.NaN``. Thus, I left the for loop as is so far.

**Comment by michaelgrund** [comment]
Let's start the discussion @GenericMappingTools/pygmt-maintainers 😉. 

Currently the following cases are covered:

```
import pygmt

fig = pygmt.Figure()
fig.basemap(region=[0, 10, 0, 11], projection="X10c/10c", frame=True)

# single line with only a label given, xmin and xmax are defined by the map borders
fig.hlines(1, label="line1")
# multiple lines of same style covered under one label
fig.hlines([2, 3], pen="2p,dodgerblue4", label="line2")
# multiple lines of same style covered under one label, both with same defined xmin and xmax 
fig.hlines([4, 5], xmin=2, xmax=8, pen="2p,red3", label="line3")
# multiple lines of same style covered under one label, with different defined xmin and xmax 
fig.hlines([6, 7], xmin=[1, 3], xmax=[8, 7], pen="3p,seagreen", label="line4")
# multiple lines of different style and labels as well as different defined xmin and xmax
fig.hlines([8, 9, 10], 
           xmin=[1.3, 3, 2], 
           xmax=[6.5, 7, 5], 
           pen=["4p,darkmagenta","2p,gold,--","3.5p,blue,."],
           label=["line5", "line6", "line7"])

fig.legend()
fig.show()
```
index

**Review by meghanrjones — COMMENTED** [review]
This is nice, thanks for implementing!

Am I wondering whether you intend for this function to be available for only Cartesian plots or both Cartesian and geographic plots? For the former, an error would need to be raised for projections other than -JX. For the latter, some adjustments are needed to support the various projections for global maps and to document that great circles lines between constant latitudes are plotted rather than horizontal lines.

I do not understand what ``no_clip`` does in the context of hlines, since this is used to control the clipping of symbols in plot. Could you please explain?

**Comment by willschlitzer** [comment]
@michaelgrund Are you hoping to get this in to the v0.7 release?

**Comment by michaelgrund** [comment]
> @michaelgrund Are you hoping to get this in to the v0.6 release?

Unfortunately not @willschlitzer ... Hope I can continue working on this PR end of February/beginning of March. Set the milestone to v0.7.0.

**Comment by seisman** [comment]
@michaelgrund I think we should get this feature done in v0.13.0. Are you still interested in working on this PR?

File: pygmt/src/hlines.py
Comment (seisman) [inline pygmt/src/hlines.py:80]: Use the `Figure.region` property (https://www.pygmt.org/dev/api/generated/pygmt.Figure.html#pygmt-figure):
```suggestion
            # get limits from current map boundings if not given via xmin, xmax
            xmin, xmax = fig.region[0:2]
            x = np.repeat([[xmin], [xmax]], list_length, axis=1)
```

**Comment by seisman** [comment]
Ping @michaelgrund to see if you have time to finish the PR in v0.14.0.

**Comment by seisman** [comment]
I'll take over this PR and hopefully can finish it before v0.14.0

**Comment by yvonnefroehlich** [comment]
@seisman related to your comment in https://github.com/GenericMappingTools/pygmt/pull/3593#issuecomment-2475457511 should this new high-level method for specifically plotting horizontal lines be included in the "Plotting Lines" tutorial? Or do we want to have a separate gallery example showing the different interpretation of horizontal lines in different coordinate systems?

**Comment by seisman** [comment]
> Or do we want to have a separate gallery example showing the different interpretation of horizontal lines in different coordinate systems?

I prefer this one

**Comment by seisman** [comment]
The plan is to implement hlines and vlines, then add a gallery example for plotting horizontal and vertical lines in Cartesian, polar and geographic projections.

**Comment by michaelgrund** [comment]
Thanks for pushing this forward @seisman after almost 3.5 years. I'm fine with all the changes, especially with removing too complicated parts I added initially. 

> The plan is to implement hlines and vlines, then add a gallery example for plotting horizontal and vertical lines in Cartesian, polar and geographic projections.

If I remember correctly the gallery example I provided within this PR was the left one in the example figure of the first post. After implementing ``vlines`` adding this together with some geographic and polar plots might be a good gallery example.

**Review by seisman — APPROVED** [review]

File: pygmt/src/hlines.py
Comment (michaelgrund) [inline pygmt/src/hlines.py:42]:
```suggestion
    - **Polar** projection: lines are plotted as arcs along a constant radius.
```

File: pygmt/src/hlines.py
Comment (michaelgrund) [inline pygmt/src/hlines.py:115]:
```suggestion
        # 1. Only specify a label when plotting the first line.
```

File: pygmt/src/hlines.py
Comment (michaelgrund) [inline pygmt/src/hlines.py:122]:
```suggestion
        # systems and straight lines in Cartesian coordinate systems (including polar
```

File: pygmt/src/hlines.py
Comment (michaelgrund) [inline pygmt/src/hlines.py:124]:
```suggestion
        # coordinate systems) or constant radius (in polar projection), we need to
```

**Review by michaelgrund — COMMENTED** [review]
Looks good!

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers I'd like to see this feature in PyGMT v0.14.0, so please give this PR a high priority.

After approval, I'd like to implement the `Figure.vlines` too in v0.14.0. If having time, maybe add a gallery example for `hlines`/`vlines` in v0.14.0.

File: pygmt/src/hlines.py
Comment (yvonnefroehlich) [inline pygmt/src/hlines.py:62]:
```suggestion
        If ``True``, do not clip lines outside the plot region. Only makes sense in the
        Cartesian coordinate system.
```

File: pygmt/src/hlines.py
Comment (yvonnefroehlich) [inline pygmt/src/hlines.py:76]:
```suggestion
    >>> fig.hlines(y=1, pen="1p,black", label="Line at y=1")
    >>> fig.hlines(y=2, xmin=2, xmax=8, pen="1p,red,-", label="Line at y=2")
    >>> fig.hlines(y=[3, 4], xmin=3, xmax=7, pen="1p,black,.", label="Lines at y=3,4")
    >>> fig.hlines(y=[5, 6], xmin=4, xmax=9, pen="1p,red", label="Lines at y=5,6")
    >>> fig.hlines(y=[7, 8], xmin=[0, 1], xmax=[7, 8], pen="1p,blue", label="Lines at y=7,8")
```

File: pygmt/tests/test_hlines.py
Comment (yvonnefroehlich) [inline pygmt/tests/test_hlines.py:41]: Strictly speaking it's plural, but as this is "only" a test, probably OK to leave it as is.

```suggestion
    fig.hlines([11, 12], xmin=1, xmax=9, pen="1p,blue", label="Lines at y=11,12")
    fig.hlines(
        [13, 14], xmin=[3, 4], xmax=[8, 9], pen="1p,red", label="Lines at y=13,14"
    )
```

**Review by yvonnefroehlich — APPROVED** [review]

**Pull request merged**
